add Lsn type

ericseppanen opened the pull request:
This type is a zero-cost wrapper for a u64, meant to help code communicate with precision what that value means, and prevent any type mix-ups.

It implements `Display` and `Debug`. `Display` will format as `1234ABCD:5678CDEF` while `Debug` will format as `Lsn(1234567890)`.

This isn't ready to merge; it's for code review purposes only.
There are lots of FIXME comments that want feedback.

Fixes #49.

knizhnik reviewed (commented):
Well, I can not agree that it is "zero-cost wrapper".
Please look at this simple program:
```
use std::time::Instant;
use std::ops::{Add};

#[derive(Debug, Clone, Copy, Eq, Ord, PartialEq, PartialOrd)]
pub struct Lsn(pub u64);

impl From<u64> for Lsn {
    fn from(n: u64) -> Self {
        Lsn(n)
    }
}

impl Add<u64> for Lsn {
    type Output = Lsn;

    fn add(self, other: u64) -> Self::Output {
        // panic if the addition overflows.
        Lsn(self.0.checked_add(other).unwrap())
    }
}
fn inc_lsn_struct(lsn: Lsn) -> Lsn {
	lsn + 1
}

fn inc_lsn(lsn: u64) -> u64 {
	lsn + 1
}

pub fn main() {
	const ITERATIONS: u64 = 1000000000;
	{
		let now = Instant::now();
		let mut lsn = Lsn(0);
		for _ in 1..ITERATIONS {
			lsn = inc_lsn_struct(lsn);
		}
		println!("Elapsed for LSN {:?} struct {:?}", lsn, now.elapsed());
	}
	{
		let now = Instant::now();
		let mut lsn : u64 = 0;
		for _ in 1..ITERATIONS {
			lsn = inc_lsn(lsn);
		}
		println!("Elapsed for LSN {:?} struct {:?}", lsn, now.elapsed());
	}
}
```

Results at my notebooks are the following:

```
knizhnik@xps:~$ rustc -O call.rs
knizhnik@xps:~$ ./call 
Elapsed for LSN Lsn(999999999) struct 27.057152ms
Elapsed for LSN 999999999 struct 39ns
knizhnik@xps:~$ rustc call.rs
knizhnik@xps:~$ ./call 
Elapsed for LSN Lsn(999999999) struct 45.84106434s
Elapsed for LSN 999999999 struct 30.92782111s
```
It is about 1.5 time differences for non-optimized code and 1000 times for optimized.
I am not sure that optimizer doesn't eliminate calls at all in last case. And speed is still very high. I do not think that we will notice any difference if we use this Lsn representation.

But still it is not possible to say that cost is zero:)

hlinnaka commented on pageserver/src/page_cache.rs:
```diff
@@ -218,18 +219,18 @@ fn init_page_cache(conf: &PageServerConf, timelineid: ZTimelineId) -> PageCache
     pub fn pack(&self, buf: &mut BytesMut) {
         self.tag.pack(buf);
-        buf.put_u64(self.lsn);
+        buf.put_u64(self.lsn.0);
```
Why is `lsn.0` needed, why not just `lsn`? I thought the `impl From<Lsn> for u64` trait would make that work.

ericseppanen replied:
The function takes only a u64:
```
fn put_u64(&mut self, n: u64);
```

There are a few ways to get automatic conversions, but none of them happen to be used here:
- Rust always tries to do "autoref" of the `self` argument. This is what allows you to type, e.g. `myvec.len()` instead of `(&my_vec).len()`.
- Some functions have generic parameters that can handle the coercion, e.g. one could write:
```
fn put_u64<T>(&mut self, n: T)
where T: Into<u64>

{
    let n: u64 = n.into();
    // do stuff with n
}
```
- You can do a similar generic type for references using `AsRef` or `AsMut`. You'll see this in the standard library a lot, e.g. `File` has `open<P: AsRef<Path>>(path: P)`, and there's a [long list of `std` types](https://doc.rust-lang.org/std/path/struct.Path.html#impl-AsRef%3CPath%3E) that implement `AsRef<Path`.

The question mark operator (for propagating errors) uses `Into<___>` for whatever your returned error type is. So even if you're not using `anyhow` or `thiserror`, implementing `Into` (or more likely, `From`) for your error types will allow you to use a variety of error types with `?`.

ericseppanen replied:
I do agree that `buf.put_u64(self.lsn.0)` isn't very friendly.

However, I think that all of this manual byte-pushing code should be replaced with `serde` anyway. As long as we `#[derive(Serialize, Deserialize)]` on the `Lsn` type then it will work exactly the same as u64 with no extra code at all.

hlinnaka replied:
It's not a big deal, we can live with it. I was just surprised it didn't "just work".

> However, I think that all of this manual byte-pushing code should be replaced with serde anyway. As long as we #[derive(Serialize, Deserialize)] on the Lsn type then it will work exactly the same as u64 with no extra code at all.

That handles this particular case, but I would imagine that this arises in many other places, too. One example off the top of my head: what if you want to store an Lsn in an AtomicU64?

I think this works, too:
```
    buf.put_u64(self.lsn.into());
```

ericseppanen replied:
I haven't found a good crate that allows `Atomic<T>` where it's implemented as e.g. `AtomicU64` if the size is 64-bits. I made one myself, but never quite finished it. Handling the case where `T` is not `Copy` gets quite nasty, but we don't need that here, so I could probably dig out the simple version (that requires `T = Copy`) if we want it.

Or we could just do a quick `AtomicLsn` wrapper, and wait to see how often we need to cut/paste that code...

ericseppanen replied:
Another thought: the kind of argument mismatch that `put_u64` hits gets less likely as we get to more complex types (that aren't `Copy`), because they're likely to be passed by reference.

If we have a bunch of functions that take by-reference arguments, e.g.
```
fn handle_mytype(mt: &MyType);
```

Then you'll note that the `mt` argument can be satisfied by any container that `Deref`s to a MyType. For example, `Arc<MyType>` or `MutexGuard<'_, MyType>` both `impl Deref` with `Target=Mytype`, so you can write:
```
handle_mytype(&mt);
```
... and that code will continue to work whether `mt` is a `MyType` or `Arc` or `MutexGuard` (or some other container that we build ourselves).

So your instinct that wrapper types should "just work" is still partially true, at least for the bigger data structures.

knizhnik replied:
Will it be better to add pack/unpack methods to Lsn?
So that it is possible to write:
```
self.tag.pack(buf);
self.lsn.pack(buf);
```

ericseppanen replied:
> Will it be better to add pack/unpack methods to Lsn?

Yes, this can be done automatically with `#[derive(serde::Serialize, serde::Deserialize)]` and then we can use any serde-compatible library we like. The `bincode` crate seems to implement the raw serialization we need in most places (including endian conversions).

I would need a bit more time to look at the code, but the end result is that we should be able to write things like `my_data_structure.serialize(&mut buffer)` without ever needing to type `put_u64` again.

There is one case where I'm not certain about, and that's when the network packet includes little-endian data that can be copied directly into a Rust struct with no validation. I don't think it's possible to generate that code with `serde` right now (see also #6), but if the performance is important to us, I think we can build tools to do that.

hlinnaka reviewed (approved):
Overall, looks good to me.

ericseppanen commented:
> Well, I can not agree that it is "zero-cost wrapper".
> Please look at this simple program:
> 
> `<snip>`
> 
> It is about 1.5 time differences for non-optimized code and 1000 times for optimized.
> I am not sure that optimizer doesn't eliminate calls at all in last case. And speed is still very high. I do not think that we will notice any difference if we use this Lsn representation.
> 
> But still it is not possible to say that cost is zero:)

Ah, but you cheated :)

The `Lsn` version is using `checked_add` to protect against overflow.  If you change the u64 version to also use `checked_add` you will see [identical performance](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=54f15a7367330ee374cd147ab187a930).

It's true that you will see the extra layer of abstraction in a debug build, but it release mode Rust is very good at seeing right through abstractions like this.

Zero-cost abstractions are one of Rust's primary design goals. Rust optimization is a bit more powerful than C or C++, because the type system carries a lot more information. And the Rust optimizer will never break your code the way that often happens to C/C++ programs that didn't know they were relying on some hidden undefined behavior.

ericseppanen commented:
And one of these days, rust optimization will get even better! There has been a long fight against LLVM bugs, but in March the nightly compiler [enabled](https://github.com/rust-lang/rust/pull/82834) the "mutable noalias" optimization, which means that code like this can be optimized (because mutable references are unique, these two references are known to not point to the same location):
```
fn adds(a: &mut i32, b: &mut i32) {
    *a += *b;
    *a += *b;
}
```
This optimization has been enabled and reverted a few times (because of bugs in LLVM), so we might not see it in a stable release yet. But it will happen eventually.

ericseppanen commented on zenith_utils/src/lsn.rs:
```diff
@@ -0,0 +1,138 @@
+impl fmt::Display for Lsn {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // FIXME: should this be "{:X}/{:08X}" ?
+        write!(f, "{:X}/{:X}", self.0 >> 32, self.0 & 0xffffffff)
```
I'll defer to the people with more postgres knowledge. Which one would you like?

ericseppanen commented on zenith_utils/src/lsn.rs:
```diff
@@ -0,0 +1,138 @@
+    type Output = Lsn;
+
+    fn sub(self, other: u64) -> Self::Output {
+        // panic if the subtraction overflows/underflows.
```
We could also drop `sub` and instead implement `checked_sub`, returning `Option<Lsn>`, forcing the caller to choose a strategy.

In the standard library, there is also [`saturating_sub`](https://doc.rust-lang.org/std/primitive.u64.html#method.saturating_sub), which returns the min or max instead of overflowing.

ericseppanen commented on zenith_utils/src/lsn.rs:
```diff
@@ -0,0 +1,138 @@
+
+    #[test]
+    fn test_lsn_strings() {
+        assert_eq!("12345678/AAAA5555".parse(), Ok(Lsn(0x12345678AAAA5555)));
```
Thanks! I should test for `"-1234/-5789"` etc.

ericseppanen commented:
Updates:
- rebased the branch
- replaced `sub` with `sub_checked` ([bbfa4fb](https://github.com/zenithdb/zenith/pull/64/commits/bbfa4fbe6b7091cf7e821ae56928c94aae9597d3))
- added an `AtomicLsn` type, to reduce some of the manual conversions ([84fe905](https://github.com/zenithdb/zenith/pull/64/commits/84fe9051f92d641d3310200dcc878d1b89e755f0))

ericseppanen commented on pageserver/src/page_cache.rs:
```diff
@@ -95,9 +96,12 @@ impl AddAssign for PageCacheStats {
-            last_valid_lsn: self.last_valid_lsn + other.last_valid_lsn,
-            last_record_lsn: self.last_record_lsn + other.last_record_lsn,
+
+            // FIXME: needs review
+            // What should be happening here? I'm not sure what is the desired result.
+            first_valid_lsn: min(self.first_valid_lsn, other.first_valid_lsn),
+            last_valid_lsn: max(self.last_valid_lsn, other.last_valid_lsn),
+            last_record_lsn: max(self.last_record_lsn, other.last_record_lsn),
```
I think I need advice here. I'm pretty sure that adding LSNs doesn't make sense, but is the min/max stuff here any better, or should the 3 LSNs just be removed from `PageCacheStas` altogether?

knizhnik replied:
I think @hlinnaka can tell more about it: is there any sense in calculating min/max LSNs for different timelines?
There is definitely no sense in comparing LSNs from different tenants and I hope we will resurrect multi-tenancy support sometimes.

hlinnaka replied:
No, it doesn't really make sense. We can just remove them from the stats, at least from the aggregated stats. It would be nice to look at the LSNs of a specific timeline, but we don't have an UI for that so that that's academic at the moment.

ericseppanen commented on pageserver/src/waldecoder.rs:
```diff
@@ -110,15 +111,16 @@ impl WalStreamDecoder {
                 self.lsn += SizeOfXLogLongPHD as u64;
                 continue;
-            } else if self.lsn % (XLOG_BLCKSZ as u64) == 0 {
+            } else if self.lsn.0 % (XLOG_BLCKSZ as u64) == 0 {
```
I could use some help here.

What does `XLOG_BLCKSZ` mean?
What does `lsn % XLOG_BLCKSZ` mean?
Should this be a member function, e.g. `Lsn::block_offset(block_size: u64)`?
If so, is block size a compile-time constant? Is it always a power of 2?

knizhnik replied:
XLOG_BLCKSZ is mostly the same as BLCKSZ: size of page in Postgres. Postgres is mostly using bock IO for accessing files. lsn % XLOG_BLCKSZ just mean offset in the block. BLCKSZ and XLOG_BLCKSZ is compile time constants, chaging them requires recompilation of postgres and ... although it is possible, nobody is doing. SO them can be considered as  "world constants" as PI. Yes, them should be power of two.

ericseppanen commented on pageserver/src/waldecoder.rs:
```diff
@@ -184,16 +187,13 @@ impl WalStreamDecoder {
-                        );
-                        self.padlen = (WAL_SEGMENT_SIZE - (self.lsn % WAL_SEGMENT_SIZE)) as u32;
+                        trace!("saw xlog switch record at {}", self.lsn);
+                        self.padlen = (WAL_SEGMENT_SIZE - (self.lsn.0 % WAL_SEGMENT_SIZE)) as u32;
                     }
 
-                    if self.lsn % 8 != 0 {
-                        self.padlen = 8 - (self.lsn % 8) as u32;
+                    // FIXME: what does this code do?
+                    if self.lsn.0 % 8 != 0 {
+                        self.padlen = 8 - (self.lsn.0 % 8) as u32;
                     }
```
Could someone educate me about what's happening here? It looks like:
- If "switch record" (what's that?), then compute padding to the end of the segment.
- If the current LSN isn't a multiple of 8, then compute padding to the next 8-byte boundary, overwriting the padding computed above (?!)

How much of this should be captured into `Lsn` member functions? We could add something like `Lsn::pad_to(chunk_size: u64)` that could be used for both.

knizhnik replied:
WAL in postgres is splitted into segments. Default size is 16Mb, but unlike XLOGBCLSZ is can be easily changed.
Switching XLOG means start writing to the next segment file.
Usually XLOG seitch record is not needed: WAL segments is filled till the end and then we swutch to next segment. But in some cases we may need to force switching to the new segment although there is still space in the current segment. It may be needed for WAL archiving.

WAL records are aligned on 8 byte boundary. Padding is used to enforce such alignment.
Actually the code fragment can be simplified as:
```
self.padlen  = (-self.lsn.0  & 7) as u32;
```

hlinnaka replied:
If you want to see an xlog switch in action, you can force it with `select pg_switch_wal();`. Can become handy in testing. See https://www.postgresql.org/docs/10/functions-admin.html.

ericseppanen replied:
Is it a bug that the code overwrites the "switch record" padding with the 8-byte wal record padding?

ericseppanen commented on zenith_utils/src/lsn.rs:
```diff
@@ -0,0 +1,175 @@
+use std::str::FromStr;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+// FIXME: should we implement conversions to/from PgLsn?
+// Doing so would mean zenith_utils gets a dependency on postgres-types.
+// That's not necessarily a bad thing, but if there is only one place
+// that needs it, it might be easier to do it there, e.g.
+//     let x = Lsn::from(u64::from(pglsn))
```
I'm probably going to answer my own question here and say "no, let's not bother". There are very few places we need to interface with `postgres_types::PgLsn`, so I don't think this matters.

If I'm wrong, it's easy to fix later.

hlinnaka replied:
Agreed

ericseppanen commented on zenith_utils/src/lsn.rs:
```diff
@@ -0,0 +1,175 @@
+        Ok(Lsn(n))
+    }
+
+    // FIXME: Is there a more ergonomic way to do this?
+    //
+    // For example, we could have a SegmentGeometry struct that
+    // implements this logic instead, e.g.
+    //
+    //    const seg_geo = SegmentGeometry::new(16 * 1024 * 1024);
+    //    let offset = seg_geo.seg_offset(lsn);
+    //    let seg_no = seg_geo.seg_num(lsn);
+
+    /// Compute the offset into a segment
+    pub fn segment_offset(self, seg_sz: u64) -> u64 {
+        self.0 % seg_sz
+    }
```
Does anyone have an opinion about this?

I don't see many places where segment size needs to be acted on, so I'll probably leave this as-is unless someone wants me to be more ambitious...

knizhnik replied:
IMHO "segment geometry" seems to be confusing. Actually segment size is specified during Postgres configuration and page server/wal acceptor receives this infirmation from from computation node.
